### PR TITLE
Propagate configuration/route conditions to KFService status object

### PR DESCRIPTION
**Pull request by yuzisun**
**What this PR does / why we need it**:
Populate following conditions on status object for 
- DefaultConfigurationReady
- CanaryConfigurationReady
- RouteReady
- KFServiceReady

KFServiceReady condition is depending on `DefaultConfiguationReady` and `RouteReady` conditions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubeflow/kfserving/issues/113

**Special notes for your reviewer**:

1. This PR replaced our conditions definition with knative's status duck type(https://github.com/knative/pkg/blob/master/apis/duck/v1beta1/status_types.go#L54), so that we can reuse all the conditions related utility functions and easier to propagate the conditions from knative resources. The conditions definition is mostly same as what we had before, one difference is that knative uses `VolatileTime` in place of `metav1.Time` to exclude this from creating `equality.Semantic` difference when comparing status object, the side effect of this is that when  updating status object I'am getting an error `status.conditions.lastTransitionTime in body must be of type object: "string"` because `lastTransitionTime` is now of type `object`. I have chatted with kube builder guys, they suggested using marker `+kubebuilder:validation:Type=string` which is supported in `controller-runtime` 2.0,  but we are still on 1.9.0 due to knative dep. For now I have walked around the issue by patching the generated crd with string type for `lastTransitionTime`.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/145)

### CODE REVIEW

**Comment by yuzisun** [comment]
@ellis-bigelow @cliveseldon @rakelkar This is ready to review.

File: config/default/manager/manager.yaml [hunk 1]
```diff
@@ -34,10 +34,10 @@ spec:
         resources:
           limits:
             cpu: 100m
-            memory: 30Mi
+            memory: 300Mi
```
Comment (ellis-bigelow) [inline config/default/manager/manager.yaml]: What's going on here? Why do we need more mem?

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 2]
```diff
@@ -14,41 +14,82 @@ limitations under the License.
...
-	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"github.com/knative/pkg/apis"
 	knservingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"k8s.io/api/core/v1"
 )
 
 // ConditionType represents a Service condition value
 const (
-	// ServiceConditionRoutesReady is set when the service's underlying
+	// KFServiceConditionRoutesReady is set when the service's underlying
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: I worry a little about exposing the Knative "Route" primitive in our interface. I think using the terminology "Route" is fine though. Do you mind rewording the comment to not expose the detail? Maybe something like, "once network configuration has completed"

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 3]
```diff
@@ -14,41 +14,82 @@ limitations under the License.
...
 	// routes have reported readiness.
-	ServiceConditionRoutesReady duckv1alpha1.ConditionType = "RoutesReady"
-	// ServiceConditionDefaultConfigurationsReady is set when the service's underlying
+	KFServiceConditionRoutesReady apis.ConditionType = "RoutesReady"
+	// KFServiceConditionDefaultConfigurationsReady is set when the service's underlying
 	// default configuration have reported readiness.
-	ServiceConditionDefaultConfigurationsReady duckv1alpha1.ConditionType = "DefaultConfigurationReady"
+	KFServiceConditionDefaultConfigurationsReady apis.ConditionType = "DefaultConfigurationReady"
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: I think this should just be "DefaultReady" for the same reason about exposing knative configurations through our api.

How do you feel about two for now:
"DefaultReady"
"DefaultPredictorReady"

Each of the components will need to ready up and they can rollup.

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 4]
```diff
@@ -14,41 +14,82 @@ limitations under the License.
...
 	// ServiceConditionCanaryConfigurationsReady is set when the service's underlying
 	// canary configuration have reported readiness.
-	ServiceConditionCanaryConfigurationsReady duckv1alpha1.ConditionType = "CanaryConfigurationReady"
+	KFServiceConditionCanaryConfigurationsReady apis.ConditionType = "CanaryConfigurationReady"
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: likewise.

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 5]
```diff
@@ -14,41 +14,82 @@ limitations under the License.
...
 )
 
-var serviceCondSet = duckv1alpha1.NewLivingConditionSet(
-	ServiceConditionDefaultConfigurationsReady,
-	ServiceConditionCanaryConfigurationsReady,
-	ServiceConditionRoutesReady,
+// KFService Ready condition is depending on default configuration and route readiness condition
+// canary configuration readiness condition only present when canary is used and currently does
+// not affect KFService readiness condition.
+var serviceCondSet = apis.NewLivingConditionSet(
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: mind spelling out condSet?

K8s status conditions as a list instead of a set always bothered me. I'm glad this is modeled as a set now,

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 6]
```diff
@@ -14,41 +14,82 @@ limitations under the License.
...
+func (ss *KFServiceStatus) GetCondition(t apis.ConditionType) *apis.Condition {
+	return serviceCondSet.Manage(ss).GetCondition(t)
+}
+
 // PropagateDefaultConfigurationStatus propagates the default Configuration status and applies its values
 // to the Service status.
 func (ss *KFServiceStatus) PropagateDefaultConfigurationStatus(dcs *knservingv1alpha1.ConfigurationStatus) {
 	ss.Default.Name = dcs.LatestCreatedRevisionName
-	//TODO @yuzisun populate configuration status conditions
+	cc := dcs.GetCondition(knservingv1alpha1.ConfigurationConditionReady)
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: I'm finding this variable naming of cc, dcs, etc to be really confusing to read. Are you married to it?

File: pkg/apis/serving/v1alpha1/kfservice_status_test.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+import (
+	"github.com/knative/pkg/apis/duck"
+	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"k8s.io/api/core/v1"
+	"testing"
+)
+
+func TestKFServiceDuckType(t *testing.T) {
+	tests := []struct {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status_test.go]: Seems inconsistent naming:

s/tests/cases?

File: pkg/apis/serving/v1alpha1/kfservice_status_test.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+		name string
+		t    duck.Implementable
+	}{{
+		name: "conditions",
+		t:    &duckv1beta1.Conditions{},
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status_test.go]: Is this our new testing pattern? I kinda like it, should we use it below?

File: pkg/controller/kfservice/kfservice_controller.go [hunk 9]
```diff
@@ -155,10 +155,12 @@ func (r *ReconcileService) Reconcile(request reconcile.Request) (reconcile.Resul
 	configurationReconciler := knative.NewConfigurationReconciler(r.Client)
 	routeReconciler := knative.NewRouteReconciler(r.Client)
 
+	// Initialize conditions
+	kfsvc.Status.InitializeConditions()
```
Comment (ellis-bigelow) [inline pkg/controller/kfservice/kfservice_controller.go]: What exactly is this doing? Setting them all to unknown? Why is this necessary to setup?

File: pkg/controller/kfservice/kfservice_controller.go [hunk 10]
```diff
@@ -155,10 +155,12 @@ func (r *ReconcileService) Reconcile(request reconcile.Request) (reconcile.Resul
...
 	// Reconcile configurations
 	desiredDefault := resourceBuilder.CreateKnativeConfiguration(constants.DefaultConfigurationName(kfsvc.Name),
 		kfsvc.ObjectMeta, &kfsvc.Spec.Default)
-
+	log.Info("reconciling default configuration", "name", desiredDefault.Name)
```
Comment (ellis-bigelow) [inline pkg/controller/kfservice/kfservice_controller.go]: these logs will be lost in the refactor.

File: pkg/testing/be_sematic_equal_matcher.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	"fmt"
+	"github.com/onsi/gomega/format"
+	"github.com/onsi/gomega/types"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"reflect"
+)
+
+// Gomega does not support semantic equal so adding this here in testing util
+
+func BeSematicEqual(expected interface{}) types.GomegaMatcher {
```
Comment (ellis-bigelow) [inline pkg/testing/be_sematic_equal_matcher.go]: How about Should(SemanticallyEqual())?

Comment (ellis-bigelow) [inline pkg/testing/be_sematic_equal_matcher.go]: Very cool. Is this copied from knative?

File: config/default/manager/manager.yaml (hunk 1 above)
Comment (yuzisun) [reply config/default/manager/manager.yaml]: When I deploy to a real cluster I got OOM, I think the cache is bigger when you have more resources on the kubernetes cluster.

**Comment by yuzisun** [comment]
> Really cool Dan.
> 
> I was talking to the knative folks about knative/pkg. They were talking about how controller-runtime was not good enough to meet their needs which caused them to invent many of the utilities in knative/pkg. They described it as solving day-2 problems for writing controllers vs controller-runtime solving day-1 problems (i.e. getting something working ASAP). Ideally, this would be in a separate repo someday.
> 
> What do you think about holding off on this until the refactor is in. There's a little bit of collide, and I think that PR is probably closer.

hmm, can you point me to the PR ? I was expecting these utilities exist in `controller-runtime` and later I found them in `knative pkg`, here I am mostly using the condition utils in `knative pkg` and I think should be fairly easy to migrate later, but I'd like to take a look at their PR first to determine what should we do here.

File: pkg/testing/be_sematic_equal_matcher.go (hunk 11 above)
Comment (yuzisun) [reply pkg/testing/be_sematic_equal_matcher.go]: forgot to leave a note in comment, this is mostly from https://github.com/onsi/gomega/blob/master/matchers/be_equivalent_to_matcher.go which uses  `reflect.DeepEqual` and here I want `equality.Semantic.DeepEqual`.

Comment (yuzisun) [reply pkg/testing/be_sematic_equal_matcher.go]: this is gomega's naming convention, going to chat with them to see if it is possible to upstream this.

File: pkg/controller/kfservice/kfservice_controller.go (hunk 10 above)
Comment (yuzisun) [reply pkg/controller/kfservice/kfservice_controller.go]: merged your change now

File: pkg/apis/serving/v1alpha1/kfservice_status.go (hunk 2 above)
Comment (yuzisun) [reply pkg/apis/serving/v1alpha1/kfservice_status.go]: removed in comments.

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 12]
```diff
@@ -14,41 +14,79 @@ limitations under the License.
...
+
 // PropagateDefaultConfigurationStatus propagates the default Configuration status and applies its values
 // to the Service status.
-func (ss *KFServiceStatus) PropagateDefaultConfigurationStatus(dcs *knservingv1alpha1.ConfigurationStatus) {
-	ss.Default.Name = dcs.LatestCreatedRevisionName
-	//TODO @yuzisun populate configuration status conditions
+func (ss *KFServiceStatus) PropagateDefaultConfigurationStatus(defaultConfigurationStatus *knservingv1alpha1.ConfigurationStatus) {
+	ss.Default.Name = defaultConfigurationStatus.LatestCreatedRevisionName
+	configurationCondition := defaultConfigurationStatus.GetCondition(knservingv1alpha1.ConfigurationConditionReady)
+	if configurationCondition == nil {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: Can this go in the switch?

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 13]
```diff
@@ -14,41 +14,79 @@ limitations under the License.
...
 
 // PropagateCanaryConfigurationStatus propagates the canary Configuration status and applies its values
 // to the Service status.
-func (ss *KFServiceStatus) PropagateCanaryConfigurationStatus(ccs *knservingv1alpha1.ConfigurationStatus) {
-	ss.Canary.Name = ccs.LatestCreatedRevisionName
-	//TODO @yuzisun populate configuration status conditions
+func (ss *KFServiceStatus) PropagateCanaryConfigurationStatus(canaryConfigurationStatus *knservingv1alpha1.ConfigurationStatus) {
+	ss.Canary.Name = canaryConfigurationStatus.LatestCreatedRevisionName
+	configurationCondition := canaryConfigurationStatus.GetCondition(knservingv1alpha1.ConfigurationConditionReady)
+	if configurationCondition == nil {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: Likewise.

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 14]
```diff
@@ -65,5 +103,16 @@ func (ss *KFServiceStatus) PropagateRouteStatus(rs *knservingv1alpha1.RouteStatu
 		}
 	}
 
-	//TODO @yuzisun populate route status conditions
+	rc := rs.GetCondition(knservingv1alpha1.RouteConditionReady)
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: And here
